cmd/fed_serve.go

```
@@ -0,0 +1,213 @@
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+var currentModules = map[string]string{"registry": "", "director": "", "origin": ""}
+var cModsStringList = "registry, director, origin"
+
+func fedCheckWorking(reqUrl string, server string) error {
+	expiry := time.Now().Add(10 * time.Second)
+	success := false
+	for !(success || time.Now().After(expiry)) {
+		time.Sleep(50 * time.Millisecond)
+		req, err := http.NewRequest("GET", reqUrl, nil)
+		if err != nil {
+			return err
+		}
+		httpClient := http.Client{
+			Transport: config.GetTransport(),
+			Timeout:   50 * time.Millisecond,
+		}
+		_, err = httpClient.Do(req)
+		if err != nil {
+			log.Infoln("Failed to send request to "+server+"; likely server is not up (will retry in 50ms):", err)
+		} else {
+			success = true
+			log.Debugln(server + " server appears to be functioning")
+		}
+	}
+
+	return nil
+}
+
+func fedServeStart(cmd *cobra.Command, args []string) error {
+	moduleSlice, err := cmd.Flags().GetStringSlice("modules")
+	if err != nil {
+		return errors.Wrap(err, "Failed to load modules passed via --module flag")
+	}
+	// Faking a set of certain values
+	var moduleMap map[string]uint16 = map[string]uint16{}
+	for _, mod := range moduleSlice {
+		if _, ok := currentModules[mod]; ok {
+			switch mod {
+			case "registry":
+				port, err := cmd.Flags().GetUint16("reg-port")
+				if err != nil {
+					return errors.Wrap(err, "Failed to parse registry port")
+				}
+				moduleMap[mod] = port
+			case "director":
+				port, err := cmd.Flags().GetUint16("director-port")
+				if err != nil {
+					return errors.Wrap(err, "Failed to parse director port")
+				}
+				moduleMap[mod] = port
+			case "origin":
+				port, err := cmd.Flags().GetUint16("origin-port")
+				if err != nil {
+					return errors.Wrap(err, "Failed to parse origin port")
+				}
+				moduleMap[mod] = port
+			default:
+				moduleMap[mod] = 0 //Should never happen
```

Throw an error here instead of ignoring it.